cedar-lean/Cedar/Validation/RequestEntityValidator.lean

```diff
@@ -0,0 +1,157 @@
+import Cedar.Validation.Validator
+import Cedar.Validation.Typechecker
+namespace Cedar.Validation
+
+open Cedar.Spec
+open Cedar.Data
+
+inductive RequestValidationError where
+| typeError (loc : String)
+
+abbrev RequestValidationResult := Except RequestValidationError Unit
+
+inductive EntityValidationError where
+| typeError (loc : String)
+
+abbrev EntityValidationResult := Except EntityValidationError Unit
+
+def instanceOfBoolType (b : Bool) (bty : BoolType) : Bool :=
+match b, bty with
+  | true, .tt => true
+  | false, .ff => true
+  | _, .anyBool => true
+  | _, _ => false
+
+def instanceOfEntityType (e : EntityUID) (ety : EntityType ) : Bool := ety == e.ty
+
+def instanceOfExtType (ext : Ext) (extty: ExtType) : Bool :=
+match ext, extty with
+  | .decimal _, .decimal => true
+  | .ipaddr _, .ipAddr => true
+  | _, _ => false
+
+def requiredAttributesPresent (r : Map Attr Value) (rty : Map Attr (Qualified CedarType)) (k : Attr) :=
+match rty.find? k with
+    | .some qty => if qty.isRequired then r.contains k else true
+    | _ => true
+
+def instanceOfType (v : Value) (ty : CedarType) : Bool := match v, ty with
+| .prim (.bool b), .bool bty => instanceOfBoolType b bty
+| .prim (.int _), .int => true
+| .prim (.string _), .string => true
+| .prim (.entityUID e), .entity ety => instanceOfEntityType e ety
+| .set s, .set ty => (s.elts.attach.map (λ ⟨v, _⟩ => instanceOfType v ty)).all id
+| .record r, .record rty =>
+  r.keys.all rty.keys.contains &&
+  (r.kvs.attach₂.all (λ ⟨(k, v), _⟩ => (match rty.find? k with
+      | .some qty => instanceOfType v qty.getType
+      | _ => true))) &&
+  rty.keys.all (requiredAttributesPresent r rty)
+| .ext x, .ext xty => instanceOfExtType x xty
+| _, _ => false
+  termination_by v
+  decreasing_by
+    all_goals simp_wf
+    case _ h₁ =>
+      have := Set.sizeOf_lt_of_mem h₁
+      omega
+    case _ h₁ =>
+      cases r
+      simp only [Map.kvs] at h₁
+      simp only [Map.mk.sizeOf_spec]
+      omega
+
+def instanceOfRequestType (request : Request) (reqty : RequestType) : Bool :=
+  if instanceOfEntityType request.principal reqty.principal then
+    if request.action == reqty.action then
+      if instanceOfEntityType request.resource reqty.resource then
+        if instanceOfType request.context (.record reqty.context) then true
+        else false
+      else false
+    else false
+  else false
+/--
+For every entity in the store,
+1. The entity's type is defined in the type store.
+2. The entity's attributes match the attribute types indicated in the type store.
+3. The entity's ancestors' types are consistent with the ancestor information
+   in the type store.
+-/
+
+def instanceOfEntitySchema (entities : Entities) (ets : EntitySchema) : EntityValidationResult :=
+entities.toList.forM (λ (uid, data) => match ets.find? uid.ty with
```

Comment:
Indent the function body with respect to `def`.  Because this lambda is quite complicated, I suggest defining it separately using a `where` clause, for readability and proofs.  When proving the equivalence of this definition to the predicate version, it will be convenient to have a named function for the lambda's body and prove its properties separately:

```
def instanceOfEntitySchema (entities : Entities) (ets : EntitySchema) : EntityValidationResult :=
  entities.toList.forM λ (uid, data) => instanceOfEntityData uid data
where 
  instanceOfEntityData uid data => 
    match ets.find? uid.ty with 
    | ...
```